pkg/controllers/drift/controller.go

```diff
@@ -0,0 +1,137 @@
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (d Drift) Reconcile(ctx context.Context, req controllerruntime.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named(controllerName).With("provisioner", req.Name))
+	node := &v1.Node{}
+	if err := d.kubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting node, %w", err)
+	}
+
+	provisionerName, provisionerExists := node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	if !provisionerExists {
+		return reconcile.Result{}, nil
+	}
+
+	if drifted, ok := node.Labels[v1alpha5.DriftedLabelKey]; ok && drifted == "true" {
+		return reconcile.Result{}, nil
+	}
+
+	provisioner := &v1alpha5.Provisioner{}
+	if err := d.kubeClient.Get(ctx, types.NamespacedName{Name: provisionerName}, provisioner); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting provisioner, %w", err)
+	}
+
+	if drifted, err := d.cloudProvider.IsNodeDrifted(ctx, provisioner, *node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting drift for node, %w", err)
+	} else if drifted {
+		stored := node.DeepCopy()
+		node.Labels[v1alpha5.DriftedLabelKey] = "true"
+		if err := d.kubeClient.Patch(ctx, node, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, fmt.Errorf("patching node, %w", err)
+		}
+	}
+
+	return reconcile.Result{RequeueAfter: 30 * time.Minute}, nil
+}
+
+func (d Drift) Builder(ctx context.Context, m manager.Manager) operatorcontroller.Builder {
+	builder := controllerruntime.
+		NewControllerManagedBy(m).
+		Named(controllerName).
+		For(&v1.Node{}).
+		Watches(
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				provisioner := o.(*v1alpha5.Provisioner)
+				if provisioner.Spec.ProviderRef == nil {
+					return requests
+				}
+				// Ensure provisioner has a defined AWSNodeTemplate
+				nodeTemplate := &v1alpha1.AWSNodeTemplate{}
+				if err := d.kubeClient.Get(ctx, types.NamespacedName{Name: provisioner.Spec.ProviderRef.Name}, nodeTemplate); err != nil {
+					logging.FromContext(ctx).Errorf("getting AWSNodeTemplates when mapping drift watch events, %s", err)
+					return requests
+				}
+				nodes := &v1.NodeList{}
+				if err := d.kubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: provisioner.Name})); err != nil {
+					logging.FromContext(ctx).Errorf("listing nodes when mapping drift watch events, %s", err)
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			})).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10})
+
+	if err := m.GetFieldIndexer().IndexField(ctx, &v1alpha5.Provisioner{}, ".spec.providerRef.name", func(rawObj client.Object) []string {
```

Comment:
We shouldn't handle this field indexing here. Since this is a generic value, we can probably move this indexing over to karpenter-core

Reply:
Can you please help me figure out the right place to put that in karpenter-core ? 
I think [here](https://github.com/aws/karpenter-core/blob/main/pkg/controllers/provisioning/provisioner.go#L108), since it's related to provisioner, but not sure of the error handling.